### PR TITLE
feat: support to sign and verify of the sm2 algorithms.

### PR DESCRIPTION
**Pull request by ymj-uos**
Signed-off-by: 杨明君 <yangmingjun@uniontech.com>

### CODE REVIEW

**Comment by ymj-uos** [comment]
rm -v -f -r temp/

mkdir temp

cd temp

export TPM2TOOLS_TCTI=device:/dev/tpm0

tpm2 flushcontext -T device:/dev/tpm0 -t
tpm2 flushcontext -T device:/dev/tpm0 -l
tpm2 flushcontext -T device:/dev/tpm0 -s

# echo -n "b2e14c5c79c6df5b85f4fe7ed8db7a262b9da7e07ccb0ea9f4747b8ccda8a4f3" | xxd -r -ps > plaintext_signing.hex
echo -n "message digest" > plaintext_signing.hex

tpm2 loadexternal -Q -C n -G sm2 -u ../standard_data/SM2-PubKey-GMT0003_5-2012.pem -c key_external.ctx

# xxd -g 4 -c 32 plaintext_signing.hex
# xxd -g 4 -c 32 ../standard_data/SM2-Demo-Signed-GMT0003_5-2012.hex

tpm2 verifysignature -c key_external.ctx -g sm3_256 -m plaintext_signing.hex  -f sm2 -s ../standard_data/SM2-Demo-Signed-GMT0003_5-2012.hex

tpm2 flushcontext -T device:/dev/tpm0 -t

tpm2 loadexternal -Q -C n -G sm2 -r ../standard_data/SM2-PriKey-GMT0003_5-2012.pem -c key_pair_external.ctx

# tpm2 readpublic -c key_pair_external.ctx -o sm2.fmt.pub --format=pem
# openssl ec -noout -text -inform PEM -in sm2.fmt.pub -pubin

tpm2 sign -c key_pair_external.ctx -s sm2 -g sm3_256 -f plain -o sig.4 plaintext_signing.hex 

# need the upper versions of openssl 1.1.1d.5
openssl dgst -sm3 \
            -verify ../standard_data/SM2-PubKey-GMT0003_5-2012.pem \
            -signature sig.4 \
            plaintext_signing.hex

openssl dgst -sm3 \
            -sign ../standard_data/SM2-PriKey-GMT0003_5-2012.pem \
            -out sig.5 \
            plaintext_signing.hex

tpm2 verifysignature -c key_pair_external.ctx -g sm3_256 -m plaintext_signing.hex -f sm2 -s sig.5
tpm2 flushcontext -T device:/dev/tpm0 -t

tpm2 createprimary -Q -C e -g sm3_256 -G sm2 -c primary.ctx
tpm2 create -Q -g sm3_256 -G sm2 -u key-pair.pub -r key-pair.priv -C primary.ctx
tpm2 flushcontext -T device:/dev/tpm0 -t

tpm2 load -Q -C primary.ctx -u key-pair.pub -r key-pair.priv -n name.load.B1_B8 -c key-pair.ctx
tpm2 readpublic -Q -c key-pair.ctx -o tpm2.sm2.pub.pem --format=pem
# openssl ec -noout -text -inform PEM -in tpm2.sm2.pub.pem -pubin
tpm2 flushcontext -T device:/dev/tpm0 -t

tpm2 sign -c key-pair.ctx -s sm2 -g sm3_256 -f plain -o sig.6 plaintext_signing.hex
openssl dgst -sm3 \
            -verify tpm2.sm2.pub.pem \
            -signature sig.6 \
            plaintext_signing.hex

chmod a+r -R .
chown uos:uos -R .

ls -l
exit 0

I use this script to check the result of TPM calculate. all of which are correct.

**Comment by ymj-uos** [comment]
[Uploading GMT.0003.2-2012.SM2-en.pdf.pdf…]()
[Uploading GMT.0003.5-2012.SM2-en.pdf.pdf…]()

**Comment by ymj-uos** [comment]
image

@idesai 《TCG_TPM2_r1p59_Part1_Architecture_pub》

**Review by idesai — CHANGES_REQUESTED** [review]

File: lib/tpm2_alg_util.c [hunk 1]
```diff
@@ -1142,3 +1163,81 @@ TPM2_ALG_ID tpm2_alg_util_get_name_alg(ESYS_CONTEXT *ectx, ESYS_TR handle) {
...
+    BYTE gy[] = {0xbc,0x37,0x36,0xa2,0xf4,0xf6,0x77,0x9c,0x59,0xbd,0xce,0xe3,0x6b,0x69,0x21,0x53,
+                0xd0,0xa9,0x87,0x7c,0xc6,0x2a,0x47,0x40,0x02,0xdf,0x32,0xe5,0x21,0x39,0xf0,0xa0};
+
+    rc = tpm2_readpublic(ectx, tr_handle, &public, NULL, NULL);
+    if(rc != tool_rc_success || NULL == public) {
+        LOG_ERR("a error to read public key from tpm device!");
+        return NULL;
+    }
+
+    size = 2 + sizeof(id) + sizeof(a) + sizeof(b) + sizeof(gx) + sizeof(gy) +
```
[the remark this answers is not in the document]
Comment (ymj-uos) [reply lib/tpm2_alg_util.c:1194]: 2 is size of variable 'ebytes'

File: lib/tpm2_alg_util.c [hunk 2]
```diff
@@ -1142,3 +1163,78 @@ TPM2_ALG_ID tpm2_alg_util_get_name_alg(ESYS_CONTEXT *ectx, ESYS_TR handle) {
     Esys_Free(name);
     return name_alg;
 }
+
+TPM2B_DIGEST *tpm2_alg_util_sm2_calculate_z_digest(ESYS_CONTEXT *ectx, ESYS_TR tr_handle) {
+
+    tool_rc rc = tool_rc_success;
+    TPM2B_PUBLIC *public = 0;
+    rc = tpm2_readpublic(ectx, tr_handle, &public, 0, 0);
```
Comment (idesai) [inline lib/tpm2_alg_util.c:1171]:
```suggestion
     TPM2B_PUBLIC *public = 0;
     tool_rc rc = tpm2_readpublic(ectx, tr_handle, &public, 0, 0);
```

**Pull request merged**